File: Objective-C/HeaderScroll/MasterViewController.m
//
//  MasterViewController.m
//  HeaderScroll
//
//  Created by Michael Nachbaur on 5/4/20.
//  Copyright © 2020 Michael Nachbaur. All rights reserved.
//

#import "MasterViewController.h"
#import "DetailViewController.h"
#import "HeaderView.h"

@interface MasterViewController ()

@property (nonatomic, strong) HeaderView *headerView;

@end

@implementation MasterViewController

- (HeaderView *)headerView {
    if (_headerView == nil) {
        UINib *nib = [UINib nibWithNibName:@"HeaderView" bundle:nil];
        _headerView = [nib instantiateWithOwner:self options:nil].firstObject;
    }
    return _headerView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.headerView.autoresizingMask = (UIViewAutoresizingFlexibleWidth |
                                        UIViewAutoresizingFlexibleHeight);
    self.headerView.titleLabel.text = nil;// @"Building a stretching UITableView header";
    self.headerView.scrollView = self.tableView;
    self.headerView.frame = CGRectMake(0,
                                       self.tableView.safeAreaInsets.top,
                                       CGRectGetWidth(self.tableView.frame),
                                       250);

    self.tableView.backgroundView = UIView.new;
    [self.tableView.backgroundView addSubview:self.headerView];
    self.tableView.contentInset = UIEdgeInsetsMake(250, 0, 0, 0);
}

- (void)viewSafeAreaInsetsDidChange {
    [super viewSafeAreaInsetsDidChange];

    self.tableView.contentInset = UIEdgeInsetsMake(250 + self.tableView.safeAreaInsets.top, 0, 0, 0);
    [self.headerView updatePosition];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    
    [self.headerView updatePosition];
}

#pragma mark - UIScrollViewDelegate methods

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    [self.headerView updatePosition];
}

#pragma mark - UITableViewDataSource methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 100;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"Cell %ld", indexPath.row];
    return cell;
}

@end

File: Swift/HeaderScroll/MasterViewController.swift
//
//  MasterViewController.swift
//  HeaderScroll
//
//  Created by Michael Nachbaur on 4/13/20.
//  Copyright © 2020 Michael Nachbaur. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController {
    var headerView: HeaderView = {
        let nib = UINib(nibName: "HeaderView", bundle: nil)
        return nib.instantiate(withOwner: self, options: nil).first as! HeaderView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        headerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        headerView.titleLabel.text = "The title of the header"
        headerView.scrollView = tableView
        headerView.frame = CGRect(
            x: 0,
            y: tableView.safeAreaInsets.top,
            width: view.frame.width,
            height: 250)

        tableView.backgroundView = UIView()
        tableView.backgroundView?.addSubview(headerView)
        tableView.contentInset = UIEdgeInsets(
            top: 250,
            left: 0,
            bottom: 0,
            right: 0)
    }
    
    override func viewSafeAreaInsetsDidChange() {
        super.viewSafeAreaInsetsDidChange()

        tableView.contentInset = UIEdgeInsets(top: 250 + tableView.safeAreaInsets.top,
                                              left: 0,
                                              bottom: 0,
                                              right: 0)
        headerView.updatePosition()
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        headerView.updatePosition()
    }
    
    // MARK: - UIScrollViewDelegate methods

    override func scrollViewDidScroll(_ scrollView: UIScrollView) {
        headerView.updatePosition()
    }

    // MARK: - UITableViewDataSource methods

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 100
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel!.text = "Cell \(indexPath.row)"
        return cell
    }
}

